File: levitate/change-events/python/change_event.py
import os
import requests
import json

LEVITATE_REFRESH_TOKEN = os.getenv("LEVITATE_REFRESH_TOKEN")
LEVITATE_ACCESS_TOKEN_API = "https://app.last9.io/api/v4/oauth/access_token"
LEVITATE_ORG = os.getenv("LEVITATE_ORG")

def api_token():
    headers = {'Content-Type': 'application/json'}
    data = {"refresh_token": LEVITATE_REFRESH_TOKEN}
    response = requests.post(LEVITATE_ACCESS_TOKEN_API, headers=headers, data=json.dumps(data))

    if response.status_code != 200:
        raise Exception("Access Token can't be retrieved, {}".format(response.text))

    body = response.json()
    return body["access_token"]

if not LEVITATE_REFRESH_TOKEN:
    raise Exception("LEVITATE_REFRESH_TOKEN is not set")

if not LEVITATE_ORG:
    raise Exception("LEVITATE_ORG is not set")

url = "https://app.last9.io/api/v4/organizations/{}/change_events".format(LEVITATE_ORG)
headers = {
    'Content-Type': 'application/json',
    'X-LAST9-API-TOKEN': 'Bearer {}'.format(api_token())
}
data = {
    "event_name": "deployment",
    "event_state": "start",
    "attributes": {
        "revision": "<git_commit_sha>",
        "user": "john.smith",
        "environment": "production"
    }
}
response = requests.put(url, headers=headers, data=json.dumps(data))

if response.status_code == 200:
    print("Change event registered successfully")
else:
    raise Exception("Change Event -> Levitate failed. {}".format(response.text))

File: levitate/change-events/ruby/change-event.rb
# frozen_string_literal: true

require "net/http"
require "uri"
require "json"

# Follow the guide on
LEVITATE_REFRESH_TOKEN = ENV["LEVITATE_REFRESH_TOKEN"]
LEVITATE_ACCESS_TOKEN_API = "https://app.last9.io/api/v4/oauth/access_token"
LEVITATE_ORG = ENV["LEVITATE_ORG"]

class Levitate
  def api_token
    uri = URI.parse(LEVITATE_ACCESS_TOKEN_API)
    request = Net::HTTP::Post.new(uri)
    request.content_type = "application/json"
    request.body = JSON.dump({
                               "refresh_token": LEVITATE_REFRESH_TOKEN
                             })
    req_options = {
      use_ssl: uri.scheme == "https",
    }

    response = Net::HTTP.start(uri.hostname, uri.port, req_options) do |http|
      http.request(request)
    end


    body = JSON.parse(response.body)
    if body["access_token"] == nil
      raise "Access Token can't be rertieved, #{body}"
    end

    body["access_token"]
  end
end


raise "LEVITATE_REFRESH_TOKEN is not set" unless LEVITATE_REFRESH_TOKEN
raise "LEVITATE_ORG is not set" unless LEVITATE_ORG

uri = URI.parse("https://app.last9.io/api/v4/organizations/#{LEVITATE_ORG}/change_events")
request = Net::HTTP::Put.new(uri)
request.content_type = "application/json"
request["X-LAST9-API-TOKEN"] = "Bearer #{Levitate.new.api_token}"
request.body = JSON.dump({
                           "event_name" => "deployment",
                           "event_state" => "start",
                           "attributes" => {
                            # Add more metadata about which service is getting deployed
                             "revision" => "<git_commit_sha>",
                             "user" => "john.smith",
                             "environment" => "production"
                           }
                         })
req_options = {
  use_ssl: uri.scheme == "https",
}

response = Net::HTTP.start(uri.hostname, uri.port, req_options) do |http|
  http.request(request)
end

if response.code == '200'
  puts "Change event registered succesfully"
else
  raise "Change Event -> Levitate failed. #{response.body}"
end
